### PR TITLE
Introduce a NONE hash to distinguish from UNKNOWN

### PR DESCRIPTION
**Pull request by sschuberth**
Previously, UNKNOWN was used for two cases:

1) An existing hash value of an unknown hash algorithm.
2) A non-existing hash value of no hash algorithm.

However, for case 1) we can still compare actual and expected hash
values by string even without knowing the algorithm. Because of this it
makes sense to distinguish these two cases by introducing NONE for case
2) and keeping UNKNOWN for case 1).

Also fix test data accordingly so that only empty hash values use NONE.

Signed-off-by: Sebastian Schuberth <sebastian.schuberth@here.com>

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/1509)

### CODE REVIEW

File: model/src/main/kotlin/Hash.kt [hunk 1]
```diff
@@ -40,6 +40,11 @@ data class Hash(
     val value: String
 ) {
     companion object {
+        /**
+         * A constant to specify no hash at all.
+         */
+        val NONE = Hash(HashAlgorithm.NONE, HashAlgorithm.NONE.toString())
+
         /**
          * A constant to specify an unknown hash.
```
Comment (fviernau) [inline model/src/main/kotlin/Hash.kt]: Is this now the constant to specify a `known` hash for an `unknown` hash algorithm?
I find it confusing that a constant for this is defined as the hash value itself can be arbitrary: is this constant used and or how about removing it?

Comment (sschuberth) [reply model/src/main/kotlin/Hash.kt]: > Is this now the constant to specify a `known` hash for an `unknown` hash algorithm?

No, this constant would be used to say "this thing does not have a hash at all". For an existign hash value generated by an unknown hash algorithm the `UNKNOWN` constant would be continued to used.

> I find it confusing that a constant for this is defined as the hash value itself can be arbitrary: is this constant used and or how about removing it?

It cannot be arbitrary. For `NONE` the hash value is always blank (which includes it being empty). So far the constant is not used, but is it added for consistency with the `UNKNOWN` constant. I plan to use the `NONE` constant later.

Comment (fviernau) [reply model/src/main/kotlin/Hash.kt]: Sorry for confusion. I was too unclear, my comment was actually referring to the `UNKNOWN` constant.

Comment (sschuberth) [reply model/src/main/kotlin/Hash.kt]: So, then yes, `UNKNOWN` is used to "specify a known hash for an unknown hash algorithm". It cannot be the other way around anyway if you think about it: If the hash value (not the algorithm) was unknown, then we would have no value, and if we have no value, we also do not know with algorithm it belongs to.

Also, in my wording I distinguish between "hash value", "hash algorithm" and "hash" (which is the combination of the former two). I'll try to make this clearer in the docs.

File: model/src/main/kotlin/Hash.kt [hunk 2]
```diff
@@ -63,12 +68,20 @@ data class Hash(
         }
     }
 
+    private val invalidAlgorithms = setOf(HashAlgorithm.NONE, HashAlgorithm.UNKNOWN)
+    private val validAlgorithms = enumValues<HashAlgorithm>().toSet() - invalidAlgorithms
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/Hash.kt]: What about `unverifiableAlgorithms` and `verifiableAlgorithms` instead?

File: model/src/main/kotlin/Provenance.kt [hunk 3]
```diff
@@ -76,7 +76,7 @@ data class Provenance(
         if (sourceArtifact != null) {
             // If the (non-empty) hashes match, we do not need to compare the other properties like the URL. Only
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/Provenance.kt]: Should be "non-blank" now.

File: model/src/main/kotlin/Provenance.kt [hunk 4]
```diff
@@ -76,7 +76,7 @@ data class Provenance(
...
             // support this for a known algorithm whose hash we were able to verify as otherwise the hash could be fake.
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/Provenance.kt]: If we accept unknown algorithms now this comment is not true anymore. Or use `canVerify` below instead of `!= NONE`.

File: model/src/main/kotlin/Hash.kt (hunk 2 above)
Comment (sschuberth) [reply model/src/main/kotlin/Hash.kt]: I was actually thinking about these, but found them to sound too awkward. But anyway, as you mention them and it better matches the verify functions, I'll make the change.

File: model/src/main/kotlin/Provenance.kt (hunk 3 above)
Comment (sschuberth) [reply model/src/main/kotlin/Provenance.kt]: Done.

File: model/src/main/kotlin/Provenance.kt (hunk 4 above)
Comment (sschuberth) [reply model/src/main/kotlin/Provenance.kt]: I've refactored the code slightly to make this check nicer to implement. I'm still not super happy myself, but I believe this is a good intermediate step for further hash refactoring I plan to do.

Comment (mnonnenmacher) [reply model/src/main/kotlin/Provenance.kt]: You could make `verifiable` a constructor argument of the enum which defaults to true, and only set it to false for `UNKNOWN` and `NONE`. Maybe that would be nicer, as the value is stored with the enum objects and not in a separate set?
